python/lancedb/table.py

```diff
@@ -524,22 +523,90 @@ def update(
 class LanceTable(Table):
     """
     A table in a LanceDB database.
+
+    This can be opened in two modes: if `version` is None then the latest version
+    is opened and can be tracked. If `version` is an integer then that version is
+    opened and this instance is fixed to that version.
     """
 
     def __init__(
-        self, connection: "lancedb.db.LanceDBConnection", name: str, version: int = None
+        self,
+        connection: "lancedb.db.LanceDBConnection",
+        name: str,
+        version: Optional[int] = None,
+        /,
+        read_consistency_interval: Optional[timedelta] = None,
     ):
         self._conn = connection
         self.name = name
-        self._version = version
+        # To be lazy, we initially store the desired version as a LanceDataset
+        self._version: Optional[int] = version
+        self._ds: Optional[LanceDataset] = version
 
-    def _reset_dataset(self, version=None):
-        try:
-            if "_dataset" in self.__dict__:
-                del self.__dict__["_dataset"]
-            self._version = version
-        except AttributeError:
-            pass
+        if self._version is not None and self.read_consistency_interval is not None:
+            raise ValueError(
+                "Cannot specify both version and read_consistency_interval."
+            )
+        self.read_consistency_interval = read_consistency_interval
+        self._last_consistency_check = None
+
+    @classmethod
+    def open(cls, db, name, **kwargs):
+        tbl = cls(db, name, **kwargs)
+        fs, path = fs_from_uri(tbl._dataset_uri)
+        file_info = fs.get_file_info(path)
+        if file_info.type != pa.fs.FileType.Directory:
+            raise FileNotFoundError(
+                f"Table {name} does not exist."
+                f"Please first call db.create_table({name}, data)"
+            )
+        register_event("open_table")
+
+        return tbl
+
+    @property
+    def _dataset_uri(self) -> str:
+        return join_uri(self._conn.uri, f"{self.name}.lance")
+
+    @property
+    def _dataset(self) -> LanceDataset:
+        # Returns the LanceDataset this wraps. This handles lazy loading (if
+        # self._ds is None) and checking for new versions (if user has specified
+        # a read_consistency_interval).
+        if not self._ds:
+            self._last_consistency_check = time.monotonic()
+            self._ds = lance.dataset(self._dataset_uri, version=self._version)
+        elif self._version is None and self.read_consistency_interval is not None:
```

Comment:
Isn't `self._version is None` implied by `self.read_consistency_interval is not None`?

Reply:
Perhaps. I really miss Rust enums :(

Reply:
Okay, I've refactor this to be cleaner. There is now a `_mode` property which either holds a `version` field (if in time-travel mode) or a `last_consistency_check` field (if in standard mode).